### PR TITLE
ENT-3756: Debug db mutex locks

### PR DESCRIPTION
**Pull request by craigcomstock**

### CODE REVIEW

**Comment by olehermanse** [comment]
@cf-bottom Could you please start a jenkins build for this PR?

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com/job/pr-pipeline/706/)

https://ci.cfengine.com/job/pr-pipeline/706/

**Comment by craigcomstock** [comment]
@vpodzime @olehermanse I guess you can only remove the WIP tag. This is ready for code review. I added a bunch of TODOs in cf-serverd related code thinking we might want to do the same thing there. So that's something for you to consider and whether or not any other binaries need this treatment for windows such as cf-monitord or cf-execd.

**Comment by craigcomstock** [comment]
I left the debug log level forcing a fflush(stdout) on each call to Log() as a separate commit thinking it might be nice as a separate PR.

File: cf-serverd/cf-serverd-functions.c [hunk 1]
```diff
@@ -222,6 +222,7 @@ GenericAgentConfig *CheckOpts(int argc, char **argv)
                 GenericAgentWriteVersion(w);
                 FileWriterDetach(w);
             }
+// TODO re-use SafeExit() here for windows
```
Comment (vpodzime) [inline cf-serverd/cf-serverd-functions.c]: So why don't we do this here? Is there any reason why we shouldn't?

File: libpromises/mutex.c [hunk 2]
```diff
@@ -51,11 +51,9 @@ int __ThreadLock(pthread_mutex_t *mutex,
 
     if (result != 0)
     {
-        /* Log() is blocking on mutexes itself inside malloc(), so maybe not
-         * the best idea here. */
-        Log(LOG_LEVEL_ERR,
-            "Locking failure at %s:%d function %s! (pthread_mutex_lock: %s)",
-            filename, lineno, funcname, GetErrorStrFromCode(result));
+        // printf instead of Log() to avoid problems with mutexes in Log()
+        printf("thread: %d, error: Locking failure at %s:%d function %s! (pthread_mutex_lock: %s)",
```
Comment (vpodzime) [inline libpromises/mutex.c]: I think this should be ``fprintf(stderr,...)``. And probably followed by ``fflush(); fsync()``.

File: libpromises/mutex.c [hunk 3]
```diff
@@ -69,11 +67,10 @@ int __ThreadUnlock(pthread_mutex_t *mutex,
...
     if (result != 0)
     {
-        /* Log() is blocking on mutexes itself inside malloc(), so maybe not
-         * the best idea here. */
-        Log(LOG_LEVEL_ERR,
-            "Locking failure at %s:%d function %s! (pthread_mutex_unlock: %s)",
-            filename, lineno, funcname, GetErrorStrFromCode(result));
+        // printf instead of Log() to avoid problems with mutexes in Log()
+        printf("thread: %d, error: Locking failure at %s:%d function %s! (pthread_mutex_unlock: %s)",
+               GetCurrentThreadId(), filename, lineno, funcname, GetErrorStrFromCode(result));
```
Comment (vpodzime) [inline libpromises/mutex.c]: Same here.

File: libutils/logging.c [hunk 4]
```diff
@@ -178,7 +178,6 @@ static void LogToConsole(const char *msg, LogLevel level, bool color)
     if (level == LOG_LEVEL_DEBUG)
     {
         fflush(stdout);
-        // TODO also possibly call sync()? Not sure if that's too "expensive" or not.
```
Comment (vpodzime) [inline libutils/logging.c]: This should not be part of this commit.

File: libutils/atexit.c [hunk 5]
```diff
@@ -71,7 +71,11 @@ void RegisterAtExitFunction(AtExitFn fn)
...
     pthread_mutex_unlock(&atexit_functions_mutex);
-#endif
-
+/*
+    Don't register atexit() functions on windows due to race conditions 
+    around lock cleanup.
+*/
+#else 
     atexit(fn);
+#endif
```
Comment (vpodzime) [inline libutils/atexit.c]: Since this change is in *libutils* I think we need to make sure all the tools use the ``SafeExit()`` mechanism, right? Otherwise the functions registered here are never run, right?

File: cf-serverd/cf-serverd-functions.c (hunk 1 above)
Comment (craigcomstock) [reply cf-serverd/cf-serverd-functions.c]: I think we should. Just wasn't my focus during the bug fixing phase.

File: libpromises/mutex.c (hunk 2 above)
Comment (craigcomstock) [reply libpromises/mutex.c]: Will change. Agreed.

File: libutils/logging.c (hunk 4 above)
Comment (craigcomstock) [reply libutils/logging.c]: Right! Will fix.

File: libutils/atexit.c (hunk 5 above)
Comment (craigcomstock) [reply libutils/atexit.c]: Yes. I will look for all use of RegisterAtExitFunction() and ensure that they also use SafeExit(). Probably move SafeExit to atexit.h/c for sharing purposes right!?

**Comment by craigcomstock** [comment]
@vpodzime after looking through the code more it seems there are many places that exit() is used which can affect our binaries. I'm wondering if the best strategy is to refactor all those calls to something like SafeExit() or to rather redefine exit() on Windows with something like:

```
#if defined(__MINGW32__)
  #define exit(x) SafeExit(x)
#endif
```

So that we don't have to refactor so much code and the change is fairly minimal and "just works" as expected.

**Comment by vpodzime** [comment]
> I'm wondering if the best strategy is to refactor all those calls to something like SafeExit() or to rather redefine exit() on Windows

I am definitely for using ``SafeExit()`` in all those places because that makes it explicit (for both the developers and the development/debugging tools). Redefining ``exit()`` could significantly complicate debugging and other contributors may find it really surprising.

**Comment by craigcomstock** [comment]
Good. I liked the clarity of SafeExit() but then wondered about catching use of exit() in new code. I suppose we could add static analysis rules maybe to guard against that if we wanted.

**Comment by craigcomstock** [comment]
after some testing might close this PR in favor of a new one with more thorough refactoring: https://github.com/cfengine/core/pull/3159

**Comment by vpodzime** [comment]
> after some testing might close this PR in favor of a new one with more thorough refactoring: #3159

OK, closing. Feel free to reopen this if it's still needed.
